backend/app/services/wiki_service.py

```
@@ -595,34 +602,61 @@ def save_generation_contents(
+            # Pages are identified by path when one is given. Normalising up front
+            # means a malformed path fails this write rather than the publish of the
+            # whole version, and keeps two spellings of one path from becoming two
+            # pages that the projection could not both honour.
+            normalized_paths: Dict[int, str] = {}
+            try:
+                for index, section in enumerate(payload.sections):
+                    if section.path:
+                        normalized_paths[index] = normalize_page_path(section.path)
+                assert_unique_within_version(normalized_paths.values())
+            except InvalidPagePath as exc:
+                raise HTTPException(status_code=400, detail=str(exc))
```

Ruff (0.16.0)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)